### ISSUE TITLE
`PwCalculation`: drop `parent_folder` validation for `nscf`/`bands`

### ISSUE THREAD

**Issue by mbercx** [aiidateam/aiida-quantumespresso#740]
When adding validation for the restart inputs in #722, we added a check that a `parent_folder` is present in the inputs for `nscf` and `bands` calculations. This makes perfect sense when checking the setup of a _separate_ `PwCalculation`, but when such a calculation is wrapped in a parent work chain that also wraps a `scf` calculation which during the logic of the work chain provides the `parent_folder`, it becomes impossible to submit such a work chain because the input validation of the `nscf` `PwCalculation` fails.

So basically we can't run the`PdosWorkChain` anymore now. 😅  I don't see a way around this, unless there is some magical validation setting that I'm not aware of?

@chrisjsewell seems I managed to break your work chain after all, without even touching it. Do I get a cookie now? 🍪

**Comment by qiaojunfeng** [comment]
This also breaks the Wannier workchain. However I found the `PwBandsWorkChain` is still functioning because in the `get_builder_from_protocol` the `calculation` is still `scf`, 
https://github.com/aiidateam/aiida-quantumespresso/blob/cb32be5361afa05bad617f00f8b187c96eb365ec/aiida_quantumespresso/workflows/pw/bands.py#L139-L149
while it is set to `bands` inside the `run_bands`
https://github.com/aiidateam/aiida-quantumespresso/blob/cb32be5361afa05bad617f00f8b187c96eb365ec/aiida_quantumespresso/workflows/pw/bands.py#L261-L262
so this bypass the input validator. 
We should also set `calculation` to `bands` in the `get_builder_from_protocol`.

For fixing this `parent_folder` issue, maybe we can remove the validation in `validate_inputs` and put it somewhere before `submit` the actual calculation inside the `PwCalculation`?

**Comment by mbercx** [comment]
Thanks for the comments @qiaojunfeng! Interesting how we indeed bypass the validator for the `PwBandsWorkChain`, probably the protocol `.yaml` file should indeed specify the `bands` calculation. Although my principle of never silently overriding inputs is violated in the `PwBandsWorkChain`, I'll leave it be for now. ^^

Note that I already have a fix (basically removing the validation) in https://github.com/aiidateam/aiida-quantumespresso/pull/741. Maybe you could review this? 🙃

**Comment by sphuber** [comment]
> For fixing this `parent_folder` issue, maybe we can remove the validation in `validate_inputs` and put it somewhere before `submit` the actual calculation inside the `PwCalculation`?

This is definitely possible and the most obvious alternative, however, this has undesirable side-effects. This means the validation won't hit until the calculation has actually been launched and it will except. For a plain `PwCalculation` this is not that big of a deal, but if it is buried deep within a work chain, it will first execute everything else of the work chain and only when it gets to the `PwCalculation` it will except and the work chain with it.

I thought about the possibility to have the wrapping workchains remove the validator, which is possible, but it would have to remove the entire validator (including the logic that are still valid and we want to keep). One would think that we can simply extract the validation of the parent folder from the validator of the entire input namespace, and add it to the `parent_folder` port, allowing us to remove it when needed, but this won't work, since the validator needs access to the `parameters` input, and so it cannot be added to the `parent_folder` port. So this direction seems a dead-end, unless we can find a way to bind the parameters input node to the validator some other way or forego the need of the parameters altogether.

**Comment by ramirezfranciscof** [comment]
I just run into this problem myself when trying to clean up my workchain (for my initial attempts I always just build in the method before submitting, so there is no exposure of inputs) and it is indeed quite a pickle.

So currently we have single input validation (insufficient for checking consistency) and an overall validator for the process. How about we allow to set a list of overall validation methods (instead of a single one) that can be individually toggled off by caller workflows? This way if a validator is not appropriate to be evaluated upfront in a workchain, it can be disabled without the caller just having to turn EVERYTHING off and needing to then do all validation themselves or risk exposure to preventable runtime errors.

BTW I would anyways run all validators when the process they apply to is being run because they probably would give clearer error message (instead of just letting the process fail). This would only disable the validation at the moment any of its parent processes is being run.

What do you think @sphuber @mbercx @qiaojunfeng ? Obviously this is a major change of aiida-core and would take some time, so in the meantime we would need to anyways implement a temporary solution for this.

**Comment by mbercx** [comment]
Thanks @ramirezfranciscof. I think what you are suggesting is close to what I'm trying to explain in [this comment](https://github.com/aiidateam/aiida-quantumespresso/pull/741#discussion_r721183953):

> I suppose we could add an` exclude_validation` argument to the `expose_inputs` method (similar to how we use `exclude` to exclude inputs). The user could pass a tuple that is very similar to the `exclude` one (i.e. using the . notation for inputs nested in P`ortNameSpace`s. One question would be how to indicate all validation, or only the top-level validation should be skipped. 

However, this will be a long term solution indeed. As I suggest in that PR, I think for now we can just leave the validation of this input up to the work chains that wrap the `PwCalculation`, as was done previously.

**Comment by sphuber** [comment]
If I understand @mbercx 's suggestion correctly, this is already possible and doesn't require new functionality. For example:
```python
class SuperWorkChain(WorkChain):

    @classmethod
    def define(cls, spec):
        super().define(spec)
        spec.expose_inputs(PwBaseWorkChain, namespace='base')
        spec.inputs['base'].validator = None
```
As you can see, exposing a workchain expose the ports with all their attributes, including the validators, which you can then remove by setting it to `None`. The problem is that this removes all the validation logic and we only want to remove part of it.

@ramirezfranciscof 's suggestion of a port allowing a list of validators to be defined would indeed allow for a more granular interface, although a list may not be the most ideal. It would be difficult for a subclasser to reference a particular validator (it would have to iterate over it and compare to `validator.__name__ == 'some-target-name'` which is a bit ugly). A dictionary may fix this. This would have to be implemented in `plumpy` where the `Port.validator` property should now accept a dict. We would have to look if this can be done in a backwards compatible way.

There _is_ something that we can do in the meantime. It is not very pretty and not the most maintainable but it would work. We can simply copy the validator of the base class and remove the part we don't need, and then use that to override the validator of the exposed input namespace. This leads to some code duplication and we have to remember to update all variants when the base validator is changed, but at least we can keep the part of the validation we need in the wrapping workchains.

**Comment by ramirezfranciscof** [comment]
> @ramirezfranciscof 's suggestion of a port allowing a list of validators to be defined would indeed allow for a more granular interface, although a list may not be the most ideal. It would be difficult for a subclasser to reference a particular validator (it would have to iterate over it and compare to `validator.__name__ == 'some-target-name'` which is a bit ugly). A dictionary may fix this. This would have to be implemented in `plumpy` where the `Port.validator` property should now accept a dict. We would have to look if this can be done in a backwards compatible way.

I guess the trivial (and most natural) way to do this is adding `Port.validators` (note the plural) as this dict of validators. Since this will already have to cycle through the dict it should be easy to just add to the cycle the one from `Port.validator` (once the overall cycling mechanism is in place). Then remove the use of `Port.validator` (singular) from all the documentation so people stop using it and some day in the future deprecate it xD.

**Comment by mbercx** [comment]
Seems I read @ramirezfranciscof's suggestion a bit too quickly. 😅  I think having such granularity would definitely be useful. 👍 

However, there is some urgency in merging #741 because of LUMI, and leaving the validation of the `parent_folder` input would mean @qiaojunfeng also has to update his work chain to fix this (and possibly others who rely on `nscf`/`bands` calculations). So I would simply remove the validation for now.

**Comment by sphuber** [comment]
> Seems I read @ramirezfranciscof's suggestion a bit too quickly. 

And this time I fear you may have read my comment too quickly :sweat_smile: Instead of removing it entirely, we should simply override it in the wrapping workchains with a clone of the validator minus the check on the parent folder.

**Comment by ramirezfranciscof** [comment]
Nice, that is quite a witty idea to deal with the problem of code duplication!

Why did you make one of them static and the other not? I mean, typically these would be outside of the class, just that here I think you did it this way to easily access it externally from the module through the entry point so that exact path inclusion is not needed (i.e. using `PwCalculation.validate_inputs_base`). Can't they be both static?

**Comment by sphuber** [comment]
The second cannot be static because it needs to call `self.validate_inputs_base`. I guess that could also change that to `PwCalculation.validate_inputs_base` but feels a bit weird if it is in a method of the class itself.

**Comment by mbercx** [comment]
This is a nice way to avoid code duplication indeed 👌. However, since it would break existing work chains, it's still a backwards-incompatible change. So maybe we should return a warning for now, that points to an explanation on how developers can update their code, and then switch to raising an error in v4.0.0?
